### PR TITLE
Also allow to increase take, add tests

### PR DESCRIPTION
**Pull request by gztensor**
Added: 

- Extrinsic `increase_take` in `subtensor` pallet
- Extrinsic `sudo_set_tx_rate_limit_delegate_take` in `admin-utils` pallet
- Separate rate limit logic for increasing delegate take (decreasing is not rate limited)
- Unit tests for decreasing and increasing take in misc. corner cases

For the code review, please check if this check is needed in `do_decrease_take`, I removed it because I think it should not be there:
https://github.com/opentensor/subtensor/blob/c1743742a82ce6643639f01527e886912789338e/pallets/subtensor/src/staking.rs#L141

This PR closes https://github.com/opentensor/subtensor/issues/314

### CODE REVIEW

File: pallets/admin-utils/tests/mock.rs [hunk 1]
```diff
@@ -84,6 +84,7 @@ parameter_types! {
     pub const InitialWeightsVersionKey: u16 = 0;
     pub const InitialServingRateLimit: u64 = 0; // No limit.
     pub const InitialTxRateLimit: u64 = 0; // Disable rate limit for testing
+    pub const InitialTxRateLimitDelegateTake: u64 = 0; // Disable rate limit for testing
```
Comment (distributedstatemachine) [inline pallets/admin-utils/tests/mock.rs]: Nit: Should we call this `InitialDelegateTakeRateLimit` ? This seems to follow the pattern of the other rate limit constants

File: pallets/subtensor/tests/staking.rs [hunk 2]
```diff
@@ -2245,3 +2245,378 @@ fn test_faucet_ok() {
...
+        // Add balance
+        SubtensorModule::add_balance_to_coldkey_account(&coldkey0, 100000);
+
+        // Register the neuron to a new network
+        let netuid = 1;
+        add_network(netuid, 0, 0);
+        register_ok_neuron(netuid, hotkey0, coldkey0, 124124);
+
+        // Coldkey / hotkey 0 become delegates with 50% take
+        assert_ok!(SubtensorModule::do_become_delegate(
```
Comment (distributedstatemachine) [inline pallets/subtensor/tests/staking.rs:2831]: How is the delegate able to set a 50% take ?

Comment (distributedstatemachine) [reply pallets/subtensor/tests/staking.rs:2831]: Get it now , it is disabled in mocks. Do we want to reenable it , as its essentiail to testing the logic ?

**Review by distributedstatemachine — CHANGES_REQUESTED** [review]
Looks good! The only thing we need to do with is PR is either rebase off development , or merge into stao branch. v1.0.0 has alot of changes and we would need to capture them here. I think it would also be nice to test the delegate take rate limits

**Review by distributedstatemachine — CHANGES_REQUESTED** [review]
Looks good! The only thing we need to do with is PR is either rebase off development , or merge into stao branch. v1.0.0 has alot of changes and we would need to capture them here. I think it would also be nice to test the delegate take rate limits

**Review by distributedstatemachine — CHANGES_REQUESTED** [review]
> Looks good! The only thing we need to do with is PR is either rebase off development , or merge into stao branch. v1.0.0 has alot of changes and we would need to capture them here. I think it would also be nice to test the delegate take rate limits  

I also think it might make sense , so rate limit at the signed extension level , so people dont bombard the chain with these calls

File: pallets/subtensor/tests/staking.rs (hunk 2 above)
Comment (gztensor) [reply pallets/subtensor/tests/staking.rs:2831]: I think it makes sense to reenable, but set the limit to 50% instead of 18% so that we don't have to rewrite the tests.

**Comment by gztensor** [comment]
1. I think the best approach is to create a feature with lowered rate limits and test it in integration tests.
2. We need to know the transaction is successful in order to enforce the rate limit. Do we know that in chain extensions?

**Comment by distributedstatemachine** [comment]
> 1. I think the best approach is to create a feature with lowered rate limits and test it in integration tests.

Ok 

> 2. We need to know the transaction is successful in order to enforce the rate limit. Do we know that in chain extensions?

You can use pre_dispatch hooks on the signed extensions to filter this out .  [Here is an example of how we use them to filter out some transactions](https://github.com/opentensor/subtensor/pull/287/files#diff-9d85ff26db441ec381d86188b389b226449eee4e9847bf0ce3301c0bb6cf93a0R1851)

File: pallets/admin-utils/tests/mock.rs (hunk 1 above)
Comment (gztensor) [reply pallets/admin-utils/tests/mock.rs]: Done

File: pallets/subtensor/tests/staking.rs (hunk 2 above)
Comment (gztensor) [reply pallets/subtensor/tests/staking.rs:2831]: Done

**Comment by gztensor** [comment]
@distributedstatemachine , I made changes you requested, and also I rebased the branch to stao (with Polkadot 1.0). Would it make sense to close this PR and open a new one to stao branch instead?

**Comment by gztensor** [comment]
Also, we could probably close [decrease_take branch](https://github.com/opentensor/subtensor/tree/decrease_take) because this branch incorporates all changes from there.

**Comment by distributedstatemachine** [comment]
> @distributedstatemachine , I made changes you requested, and also I rebased the branch to stao (with Polkadot 1.0). Would it make sense to close this PR and open a new one to stao branch instead?

Yes this makes sense 

> Also, we could probably close [decrease_take branch](https://github.com/opentensor/subtensor/tree/decrease_take) because this branch incorporates all changes from there.

Sounds good to me!
